### DISCUSSION TITLE
Attributes as constraints for generic types

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#7105] [Language Ideas]
NOTE: I could not find a duplicate on this topic while looking at the open and closed issues. If an issue/discussion already exists, please excuse my lack of attention!

_**TL;DR**_: Allow attributes to be used as constraints of generic types by declaring them using the following format: `where T: [Attribute]`.

<h1>Motivation</h1>
Attributes have always been a big part of the language. They alter how types are viewed by specific methods, or by the compiler itself. The thing is, checking for attributes has always been a task that had to be done during the runtime; see the following example:

```c#
[MyAttribute]
public class MyClass
{
    // Really important stuff goes here!
}

[...]

public void RuntimeCheck<T>() 
{
    MyAttribute attribute = (MyAttribute)Attribute.GetCustomAttribute(typeof(T), typeof(MyAttribute));
    if (attribute != null)
    {
        // Custom behaviour goes here.
    }

    // Normal behaviour goes here.
}
```
Sure, we can argument that such kinds of tests are necessary when custom behaviours are present... but what if the entire check decides wether the method continues or not? See below:

```c#
public void RuntimeCheckWithReturn<T>()
{
    MyAttribute attribute = (MyAttribute)Attribute.GetCustomAttribute(typeof(T), typeof(MyAttribute));
    if (attribute == null)
    {
        // If the attribute is not present, execution must halt and return.
        return;
    }

    // This will only execute if the attribute is present.
}
```

In this case, all that boilerplate code boils down to the known behaviour of a constraint, which could be evaluated at compile-time.

<h1>Proposal</h1>
My proposal is to allow attributes to be checked statically. The following syntax could be used to differ attribute dependencies from inheritance dependencies:

```c#
public class MyClassNoAttribute
{
    // Unsurprisingly, more important stuff goes here!
}

// Syntax in question:
public void CompileTimeCheck<T>() where T: [MyAttribute]
{
    // T has to have MyAttribute attached to it, or else this throws an error.
}

public static void Main()
{
    CompileTimeCheck<MyClass>(); // OK
    CompileTimeCheck<MyClassNoAttribute>() // Compile-time error: Type "MyClassNoAttribute" does not include the "MyAttribute" attribute.
}
```
This could allow for less boilerplate code whenever the lack of an attribute is exclusive in nature. Attributes could remain being checked during runtime if a different outcome is expected.

**Comment by RikkiGibson** [comment]
> This could allow for less boilerplate code whenever the lack of an attribute is exclusive in nature.

I wasn't sure what you meant by this. Could you clarify?

I'd be curious to know about more use cases here. It feels like the goal here is to establish guardrails. There aren't any new ways the type parameter can be used in code as a result of having this constraint, I think.

What comes to mind is that you can put attributes on type parameters, those attributes can denote that the type argument for the parameter must be a type with a certain attribute, and that requirement can be checked by a Roslyn analyzer.

**Reply by ghost** [reply]
I'd be more than happy to clarify! What I meant by that is that the lack of an attribute excludes the type from the pool of types allowed by the generic method. Assume that we have a game in which we want to save player data:
```c#
[Serializable]
public class PersonalUserData
{
	// Some kind of user data that is of lesser importance:
	// Username, Keybinds, whatever else...
}

[GZipCompress]
[Serializable]
public class CharacterData
{
	// Information that would be unfair if changed:
	// Max health/mana, Inventory, etc...
}

public class SomethingElse
{
	// Something not meant to be serialized at all. Can be anything you desire!
}

[...]

private byte[] CompressZstd(string source)
{
	// Compresses data using Zstd.
}

private byte[] CompressGZip(string source)
{
	// Compresses data using Zstd.
}

public string Serialize<T>(T source) where T : [Serializable]
{
	// Serializes types tagged as Serializable.
}

public byte[] SerializeZstd<T>(T source) where T : [Serializable, ZstdCompress]
{
	// Serializes and then compresses types using Zstd.
	string maybeJson = Serialize(source); // T is implicit.
	return CompressZstd(maybeJson);
}

public byte[] SerializeGZip<T>(T source) where T : [Serializable, GZipCompress]
{
	// Serialized and then compresses types using GZip. 
	string maybeJson = Serialize(source); // T is equally implicit.
	return CompressGZip(maybeJson);
}
```

The language itself would then warn the user of misusage of the method (as you pointed out yourself: it would serve as guardrails):

```c#
public static void Main()
{
	PersonalUserData personalData = new PersonalUserData();
	CharacterData characterData = new CharacterData();
	SomethingElse somethingElse = new SomethingElse();
	_ = Serialize(personalData); // OK
	_ = Serialize(characterData); // OK
	_ = Serialize(somethingElse); // SomethingElse is not serializable!

	_ = SerializeGZip(characterData); // OK
	_ = SerializeZstd(characterData); // CharacterData cannot be compressed using Zstd! (Possibly could cause errors down the line if left to runtime checks, since the data would be compressed using Zstd and decompressed using GZip). 
}
```

Changing the mode of compression of CharacterData to `[ZstdCompress]` would automatically allow for the second method to be used, and instantly warn about the wrong decompression method down the line, since:

```c#
[...]
	CharacterData characterData = DeserializeGZip<CharacterData(byte[] data); // Would throw an error, since CharacterData is no longer tagged as GZipCompress, but rather ZstdCompress.
[...]
```

**Comment by svick** [comment]
Wouldn't a marker interface work instead of an attribute here?

**Reply by ghost** [reply]
Assume that we want to declare a component in an ECS:

```c#
public struct MyComponent : IDependsOn<Comp1, Comp2>
{
	// Information goes here...
}
```

The code would quicky become bloated since, just like Action and Func, you'd have to declare to declare a new interface everytime you want to increase the dependency by 1 (take `Func<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, TResult>` as an extreme example). Method declarations would also become impossible since you cannot declare generic dependency as constraints as far as i know:

```c#
public void MyMethod<T>() where T : IDependsOn<C>, IDependsOn<C1, C2>, IDependsOn<C1,C2,C3> { ... } // Impossible (?)
```

On the other hand, this can be easily circumvented using an attribute that takes an array of types as arguments:

```c#
[AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class, AllowMultiple = false)]
public class ComponentDependsOnAttribute : Attribute
{
    public readonly Type[] Types;

    public ComponentDependsOnAttribute(params Type[] components)
    {
        Types = components;
    }
}

[ComponentDependsOn(typeof(PlayerComponent), typeof(TransformComponent))]
public struct CameraTargetComponent { }

[ComponentDependsOn(typeof(TransformComponent))]
public struct VelocityComponent { }

[...]

public void Add<T>()
{
	// Adds a generic component to the entity.
}

public void AddWithDependency<T>() where T : [ComponentDependsOn]
{
	// Access the type list inside the attribute and add them with dependency recursively.
}
```

**Reply by HaloFour** [reply]
If `IDependsOn<...>` all extended from the non-generic `IDependsOn` you'd end up with a single interface that you could use as the generic type constraint.  That puts you into the same exact boat as a `[ComponentDependsOn]` attribute constraint given you cannot actually use that to constraint to a specific set of parameters or even an arity of parameters unless you define a separate attribute for each combination, and you end up still negotiating that relationship using reflection at runtime anyway.

**Reply by svick** [reply]
You could also do this:

```c#
public struct MyComponent : IDependsOn<Comp1>, IDependsOn<Comp2>
```
